internal/log/slog.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2021 The prometheus-operator Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package log
+
+import (
+	"fmt"
+	"log/slog"
+	"os"
+	"strings"
+	"time"
+)
+
+// NewLoggerSlog returns a *slog.Logger that prints in the provided format at the
+// provided level with a UTC timestamp and the caller of the log entry.
+func NewLoggerSlog(c Config) (*slog.Logger, error) {
+	lvlOption, err := parseLevel(c.Level)
+	if err != nil {
+		return nil, err
+	}
+
+	logger, err := parseFmt(lvlOption, c.Format)
+	if err != nil {
+		return nil, err
+	}
+
+	return logger, nil
+}
+
+func replaceSlogAttributes(_ []string, a slog.Attr) slog.Attr {
+	if a.Key == "time" {
+		return slog.Attr{
+			Key:   "ts",
+			Value: slog.StringValue(a.Value.Time().UTC().Format(time.RFC3339Nano)),
+		}
+	}
+
+	if a.Key == "level" {
+		return slog.Attr{
+			Key:   "level",
+			Value: slog.StringValue(strings.ToLower(a.Value.String())),
+		}
+	}
+
+	if a.Key == "source" {
+		return slog.Attr{
+			Key:   "caller",
+			Value: a.Value,
+		}
+	}
+
+	return a
+}
+
+func parseFmt(lvlOption slog.Level, format string) (*slog.Logger, error) {
+	var logger *slog.Logger
+	switch strings.ToLower(format) {
+	case FormatLogFmt:
+		h := slog.NewTextHandler(os.Stdout, &slog.HandlerOptions{
+			Level:       lvlOption,
+			AddSource:   true,
+			ReplaceAttr: replaceSlogAttributes,
+		})
+		logger = slog.New(h)
+		return logger, nil
+	case FormatJSON:
+		h := slog.NewJSONHandler(os.Stdout, &slog.HandlerOptions{
+			Level:       lvlOption,
+			AddSource:   true,
+			ReplaceAttr: replaceSlogAttributes,
+		})
+		logger = slog.New(h)
+		return logger, nil
+	default:
+		return nil, fmt.Errorf("log format %s unknown, %v are possible values", format, AvailableLogFormats)
+	}
+}
+
+func parseLevel(lvl string) (slog.Level, error) {
```

Comment:
```suggestion
func parseLevel(lvl string) (*slog.Level, error) {
	level := slog.Level(1)
	err := level.UnmarshalText([]byte(lvl))
	if err != nil {
		return nil, err
	}
	return &level, nil
}
```

Wdyt this approach?

Reply:
yep. it makes sense because simplify a lot of code. 

But taking a look at the implementation, the [UnmarshalText](https://github.com/golang/go/blob/master/src/log/slog/level.go#L107-L146) just works for DEBUG, INFO, WARN and ERROR. So the other cases (ALL and None) won't be covered

What do you suggest in that case?

Reply:
IMH if we get none we should use default info and about all I don't know if I follow what this means?

If you set all, are you expecting getting, debug, info, warn, and error logs?

Log use to work like:
error -> only error logs
warn -> only warn and error logs
info -> only info, warn and error logs
debug -> all logs

wdyt?

Reply:
> If you set all, are you expecting getting, debug, info, warn, and error logs?

Today the default behavior for `ALL`, is to use `debug` as I set.

> IMH if we get none we should use default info 

Today, the default behavior is to print nothing.

Reply:
I think debug is way to verbose to be used as default, we should go as info

Reply:
So, just to confirm if I understand the level thing

LevelAll = `debug level`
LevelNone = `info level` 

that is it?

---

About the code, using the `level.UnmarshalText`. Will look like this?

```Golang
func parseLevel(lvl string) (*slog.Level, error) {
	level := slog.Level(1)
	if lvl == LevelNone {
		level = slog.LevelInfo
		return &level, nil
	}
	if lvl == LevelAll {
		level = slog.LevelDebug
		return &level, nil
	}
	err := level.UnmarshalText([]byte(lvl))
	if err != nil {
		return nil, err
	}
	return &level, nil
}
```